### ISSUE TITLE
Binary Spine format throws errors. Is it or will it be supported?

### ISSUE THREAD

**Issue by Friksel** [pixijs/pixi-spine#363]
Trying out the pretty handy binary format the Spine editor can export to in pixi-spine I get an error: `The spineData param is required.` on line 8195 of pixi-spine.js.

This binary format is pretty great. The output data-filesize is half (!!) the size, so quicker to load, especially on mobile (and as a little bonus; for newbies it's way more intimidating and less easy to steal animation content as we can give the file-extension we want and people need binary tools to view the content or have to know it's made in Spine).

Maybe I'm missing something. Do we perhaps need another loader or something? Is the binary format supported by pixi-spine? 
And if not; are there any plans on adding support for it?  I would definitely use it on all my projects!

Thanks in advance!

**Comment by ivanpopelyshev** [comment]
It works.

Go to devtools network page, look if the file actually loaded, maybe its your webserver problem.

If you dont find anything, please give me full demo to ivan.popelyshev@gmail.com in zip file

**Comment by Friksel** [comment]
> It works.
> Go to devtools network page, look if the file actually loaded, maybe its your webserver problem.

Cool! I see the file getting loaded. But there is no `spineData` inside the loaded data. So this is most problably something I am doing wrong here.

This is how I create the Spine object from the loaded resource: 
```
this.tree = new PIXI.spine.Spine(this.loader.resources.tree.spineData);
```
But the loaded resource doesn't have a `spineData` object/variable.

I see the resource object has a `data` string though. But when feeding that to the Spine constructor like this:
```
this.tree = new PIXI.spine.Spine(this.loader.resources.tree.data);
```
Pixi-spine complains that the spineData object can't be a string: `Error: spineData param cant be string. Please use spine.Spine.fromAtlas("YOUR_RESOURCE_NAME") from now on.`

I've than tried  to use `fromAtlas()` eventhough it seems illogical, because I don't want to load only the atlas. So that wasn't working either.

I've also tried
```
this.tree = new PIXI.spine.Spine(this.loader.resources.tree);
```
But than Pixi-spine cannot create a new Skeleton: `Cannot read property 'length' of undefined`


**conclusion**
So it seems like something needs to be different when importing from binary compared to the json. And I am clearly doing something wrong, as I don't know what to do differently compared to the 'normal' json-import.

How are you importing the binary format from the loaded (by pixi) resource?

**Comment by ivanpopelyshev** [comment]
Yes, you just swap json for binary and it works. Please give me whole demo

**Comment by Friksel** [comment]
> Yes, you just swap json for binary and it works. Please give me whole demo

Thanks for the quick response. I was hoping for this. Glad to be able to tick that one off the list.

Sorry, I understand why you're asking to send the whole demo, but I never send full (client)projects.

I'm seeying something new now though that could lead to the solution I think: 
when changing the file extension to `.skel` I get a different error message, that could indicate that it passed the initial issue?

Now I'm getting a message about the texture size:
`Uncaught Error: Texture Error: frame does not fit inside the base Texture dimensions: X: 2 + 227 = 229 > 1923 or Y: 186 + 880 = 1066 > 885`

**Comment by ivanpopelyshev** [comment]
There are many things that can go wrong there. You didnt mention whether you actually tried to place breakpoints in pixi-spine loader itself. 

Yes, in perfect world we would have error messages for everything in that loader. Here, we have only a few.

All I can say - it should work, it worked before. The only way to find whats wrong is to debug it. PixiJS and pixi-spine are not black boxes. Its open-source projects that can exist only because people can go inside, find out whats wrong and fix it themselves or ask us to fix it.

> but I never send full (client)projects.

Then make minimal one.

For example, just yesterday I solved this one: https://www.html5gamedevs.com/topic/46549-error-cannot-read-property-spine-of-undefined/ , guy had a problem with his web-server.

**Comment by Friksel** [comment]
Alright, before I renamed the binary file to *.skel and got a step further in the process, but had the issue with the texture.

Now I re-exported the full skeleton file together with texture atlas / png packed from the Spine editor and everything works fine now!

So the issue here was the file-extension. It needed to be `.skel` or else it didn't work.

Looking at line 8737 in the pixi-spine.js (inside dist folder) I see that pixi-spine is filtering on the `skel`-fileextension.

I was planning to put a different random file extension to the Spine data files though, to have a little prevention of newbies understanding what fileformat this is. So that these people don't understand what format it is and leave it instead of stealing animations.

Would it be possible to make the file-extension configurable or not limited at all?

**Comment by ivanpopelyshev** [comment]
Yes, its possible. We had to change atlas extension before, it looks like Skel hasto be changed too. Care to make a PR?

**Comment by ivanpopelyshev** [comment]
something-something pass special flag through metadata maybe?

**Comment by Friksel** [comment]
> Yes, its possible. We had to change atlas extension before, it looks like Skel hasto be changed too. Care to make a PR?

Sorry, I don't use Typescript, don't have a Typescript compile environment here. But I can tell you what works:
* Add a parameter to the Spine constructor to let the user define the file-extension
https://github.com/pixijs/pixi-spine/blob/master/src/Spine.ts [line 55]
* Feed that file-extension variable to the loader 
* Replace the hardcoded `'skel'` extension on the spine Loader on lines 27 and 37 by the new file-extension variable as set in the Spine constructor
https://github.com/pixijs/pixi-spine/blob/master/src/loaders.ts

It would be nicer if we wouldn't be limited to any file-extension, but looking at the code I'm not sure if that's possible because of this line in the loader
`PIXI.LoaderResource.setExtensionXhrType('bin', PIXI.LoaderResource.XHR_RESPONSE_TYPE.BUFFER);`
Not sure what that `setExtensionXhrType` does exactly and if that has consequences when not set.

**Comment by ivanpopelyshev** [comment]
that's resource-loader stuff: https://github.com/englercj/resource-loader

It describes format that middleware is getting as input.

OK, I can make those changes, sometime later. Thank you!

**Comment by Friksel** [comment]
Okay, would be nice, but no hurries! Thanks! Also for your quick responses. Love that there is this pixi-spine thing though. Very nice to be able to use Spine animations in Pixi. Thanks for all the hard work!

**Comment by ivanpopelyshev** [comment]
>  Love that there is this pixi-spine thing though

Thank you!

EsotericSoftware saved us lots of hours by maintaining high-quality generic typescript implementation: https://github.com/EsotericSoftware/spine-runtimes/tree/3.8/spine-ts . Before that it was much harder to maintain.

All dirty code belongs to spine loader which is modified by people who use pixi-spine, all docs regarding hacks go here: https://github.com/pixijs/pixi-spine/tree/master/examples . I'll add your case there too.

I dont even use it myself, it was just my first assigned work as member of pixijs team.
